### DISCUSSION TITLE
Changing the order of levels of categorical regressors

### DISCUSSION THREAD

**Discussion by eort** [lnccbrown/HSSM#261] [Q&A] [answered]
Hi all!

I am trying to set up a regression with a categorical variable, which works fine as far as I see. Unfortunately though, the default way the model is set up, puts the experimental condition in the intercept, and the baseline condition in the slope. For interpretation purposes it would be much nicer to change them around. However, I can't make seem to make this work at the moment.

From reading the `bambi` docs, the reason is that they alphabetically sort the levels of the categorical variable when constructing the model. They allow, however,[ custom orderings](https://github.com/bambinos/bambi/issues/380) by specifying the `levels` parameter in the model specification. When I try this with `hssm`, this fails with following error message: 

```
# what I do
lvl = ['cond_B', 'cond_A']
model = hssm.HSSM(data=data, model="ddm", include=[ {
            "name": "v",
            "formula": "v ~ 1 + C(drug, levels=lvl)",
            "link": "identity"
        }])

# and what I get:
KeyError: 'lvl'
```

When I try this directly in bambi, it works though:
```
# what I do
lvl = ['cond_B', 'cond_A']
model = bambi.Model(f"rt ~ 1 + C(drug, levels=lvl)", data=data)
```
So it seems like the variable `lvl` is not passed on to bambi/formulae, to change the ordering? 

 Is this the right way to approach this problem? If so, could this be a problem with hssm? 

I hope I am making sense here.

Thanks again for this great tool!

Eduard

**Comment by digicosmos86** [comment]
Hi Eduard,

Thanks for bringing this to our attention! It seems that the issue is that `lvl` is not found when the `formula` is created from within HSSM. `bmb.Model` has `extra_namespace` argument which helps it find this variable. So what you need to do is something like this:

``` python
lvl = ['cond_B', 'cond_A']
model = hssm.HSSM(data=data, model="ddm", include=[ {
            "name": "v",
            "formula": "v ~ 1 + C(drug, levels=lvl)",
            "link": "identity"
        }], extra_namespace={"lvl": lvl})
```

This helps bambi find the list that you specified the formula.

Let me know if this works :)

**Comment by frankmj** [comment, marked as answer]
Thanks Paul,
just a small edit here as I was trying something similar, and found it to
work if I do

extra_namespace= {"lvl": lvl}
instead of
extra_namespace: {"lvl": lvl}

ie just replace  ":" with "="  since this is not a dictionary entry.


M

**Comment by eort** [comment]
Thanks both of you. That does do the trick!

**Comment by yufeng-tian** [comment]
This is important for group comparisons as most psych experiments involve a mixed design and categorical variables (e.g., experimental conditions). I suggest including a tutorial specifically addressing the issue of categorical variables, OR the "Bayesian (group-difference hierarchical) test" or "estimating the impact of condition as a within-subject effect in the same model" mentioned in this discussion https://github.com/lnccbrown/HSSM/discussions/421
I am unsure if such tutorials are already on the official tutorial site, but I would like to remind you that this question is important.

**Reply by AlexanderFengler** [reply]
Hi @yufeng-tian ,

we included some parts about categorical variables in the [Main Tutorial](https://lnccbrown.github.io/HSSM/tutorials/main_tutorial/), but indeed it could be fleshed out more. 

Thank you for raising this again.

A
